Go SQL query instrumentation, traces, metrics, test-app

MattFrick opened the pull request:
Incomplete/draft PR, this is just showing progress so far: eBPF prints out the SQL query.

MattFrick commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,139 @@
+#define SPAN_ID_SIZE 8
+#define SPAN_ID_STRING_SIZE 16
+
+#define BASE_SPAN_PROPERTIES \
```
We may want to convert this to using Beyla's structure format.

MattFrick commented on bpf/headers/arguments.h:
```diff
@@ -0,0 +1,88 @@
+#include <stdbool.h>
+
+// Injected in init
+volatile const bool is_registers_abi /* TODO: set this up based on instrumented application's go version >= 1.17 */ = 1;
```
TODO: Add Go version detection of instrumented application to set this from userspace side.

MattFrick replied:
Awesome, we can remove this option and the code that uses the older ABI.

MattFrick commented on bpf/headers/go_context.h:
```diff
@@ -0,0 +1,91 @@
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} tracked_spans_by_sc SEC(".maps");
+
+static __always_inline void *get_parent_go_context(void *ctx, void *map) {
```
Taken as-is from Otel, I was wondering if we should convert to use Beyla's parent finding, e.g. find_parent_goroutine()

MattFrick commented on pkg/internal/ebpf/common/common.go:
```diff
@@ -83,6 +84,6 @@ func ReadHTTPRequestTraceAsSpan(record *ringbuf.Record) (request.Span, error) {
 		return request.Span{}, err
 	}
-
+	fmt.Println("read ReadHTTPRequestTraceAsSpan")
```
To be removed

grcevski commented:
One initial thought, we shouldn't add the pq vendor information in our main repo. You can create a go module for the testcase, so that the dependency is only there.

grcevski commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,139 @@
+// This instrumentation attaches uprobe to the following function:
+// func (db *DB) queryDC(ctx, txctx context.Context, dc *driverConn, releaseConn func(error), query string, args []any)
+SEC("uprobe/queryDC")
+int uprobe_queryDC(struct pt_regs *ctx) {
```
This looks like a simple enough probe, could we not write our own code for fetching a few fields of an object, much like the other Go uprobes we have? I feel like taking this code verbatim as is brings a lot of unnecessary other code with it, like the whole reading of context information etc. None of that is relevant to Beyla. Looks to me like 20-30 lines of code max that need to be written. WDYT?

MattFrick replied:
Updated to adapt just the core functionality, and use Beyla ring buffer.

grcevski commented on test/integration/components/sqlclient/gosqlclient.go:
```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
```
Just thinking out loud here, wouldn't it be simpler to use an in memory SQL backend for this test? The Go SQL package in vendor neutral, it might be much easier to write a test with sqlite instead of with postgres.

mariomac replied:
I agree. Sqlite is pretty simple to embed. It should be as simple as:

```go
import _ "github.com/mattn/go-sqlite3"
``` 

(or any other driver listed in the [official wiki](https://github.com/golang/go/wiki/SQLDrivers))

MattFrick replied:
I went with modernc.org/sqlite

mariomac commented on test/integration/components/sqlclient/gosqlclient.go:
```diff
@@ -0,0 +1,58 @@
+
+	// Do queries for >5 seconds (it may take Beyla that long to find this process)
+	for i := 1; i < 6; i++ {
+		rows, e := db.Query("SELECT * FROM students")
+		CheckError(e)
+		defer rows.Close()
```
I wonder, as a non-DB guy: what is a more representative trace?
- Since the `Query` function starts until it ends.
- Since the `Query` function starts until we close the `rows` object.

MattFrick replied:
Excellent points (from further above comment) about prepared statements.  Good food for thought for future improvements, to at least make spans for the DB transactions even if we can't see retrieve the query string.
I also see that queryDC takes additional arguments for placeholder values.  Since we only look at the query string, we might only find "?" placeholders.
I'm not sure what's a better representative trace, but it seems like the query.
Lots of room for future improvement, thanks for the ideas!

mariomac reviewed (commented):
As a first functionality, it looks good. But I think that, to be really useful, we need to expand the instrumentation to work with prepared statements.

The most efficient way to work with SQL is through Prepared statements, as you avoid parsing/validating SQL on each query/execution. So we should support this (presumably it's the most common way to work with SQL, as it's the most efficient). I have been digging into the code and prepared statements don't seem to use `queryDC`.

Example of prepared statement:

```go
type Users struct {
  		names *sql.Stmt
}

// statements are prepared once
func NewUsers(db *sql.DB) *Users {
	stmt, _ := db.Prepare("SELECT name FROM Users")
	return &Users{names: stmt}
}

func (u *Users) Names() []string {
	rows, _ := u.names.Query()
}
```

In the above example, there is an extra complexity for relating the address of the generated statement in the constructor with its later invocation to `Query`.

Thinking out loud: for a future improvement, we could maybe support transactions and group all the queries/execs under a given transaction as spans under a common parent ID.

grcevski commented on test/integration/docker-compose-gosql.yml:
```diff
@@ -0,0 +1,89 @@
+version: '3.8'
+
+services:
```
I'm really confused about which test files are active here. It looks like this template has Beyla, and also it's own instance of the otel-collector, jaeger etc. But then we also have similar components as part of the OATS framework. Seems to me this should really be a very simple docker compose file, only containining the sql client?

MattFrick replied:
Looks like this was accidentally included in the commits, will remove.  Good catch!

MattFrick commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,91 @@
+// Copyright The OpenTelemetry Authors
```
It is based on their code, just heavily refactored.  I'm of the understanding that we need to retain this, as it could still be considered a derivative work.  What is the proper thing to do here?

grcevski replied:
I guess I'll leave it up to your judgement then. If it's using some of their code then it should include it. After my first review of this code, I thought that writing something from scratch using our data structures would be much simpler. https://github.com/grafana/beyla/pull/298#discussion_r1330549512

MattFrick commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,91 @@
+    struct sql_request_t sql_request = {0};
+    sql_request.start_time = bpf_ktime_get_ns();
+
+    if (1 /*  TODO: HOOK UP TO CONFIG */ || should_include_db_statement) {
```
Is it safe to assume that the Go string argument will still be there intact, afterwards in the uretprobe where we send into the ringbuf?  If so, that would be a nice thing to take advantage of.

grcevski replied:
I think so, we'd capture the register values at the entry of the call. Strings are immutable in Go, so the contents won't change during the execution of the function, so I can't see how the address of the argument would change or the contents.

MattFrick replied:
After a brief test it seems to work.  I've added a buildflag in case we need to go back, e.g. if the string arg ends up on heap and GC runs.

MattFrick commented on pkg/internal/ebpf/common/spanner.go:
```diff
@@ -95,3 +101,24 @@ func extractTraceparent(traceparent [55]byte) string {
+	for i, f := range fields {
+		word := strings.ToUpper(f)
+		switch word {
+		case "FROM", "INTO", "ON":
```
Excellent suggestion.  Looking over some common SQL syntax I see some more exceptions to finding the table name.  Here's the total list so far:
ALTER TABLE table_name ADD column_name datatype
CREATE TABLE table_name
DROP INDEX table_name.index_name
TRUNCATE TABLE table_name
UPDATE table_name

grcevski replied:
Yeah those would be good to add too!

MattFrick replied:
Added.

MattFrick commented on pkg/internal/export/otel/traces.go:
```diff
@@ -347,6 +355,11 @@ func traceName(span *request.Span) string {
+		// We don't have db.name, but follow "<db.operation> <db.name>.<db.sql.table_name>"
+		operation := span.Method
+		table := span.Path
+		return operation + " ." + table
```
Excellent.  Double checking the otel semantics: "If db.sql.table is not available due to its semantics, the span SHOULD be named <db.operation> <db.name>"
I think omitting the empty space (since we don't have db.name) makes sense.
Thanks!

MattFrick commented on test/integration/components/sqlclient/gosqlite.go:
```diff
@@ -0,0 +1,57 @@
+	CheckError(e)
+	_, e = db.Exec(insertStudent, "Alice", 2)
+	CheckError(e)
+
```
I added a /sqltest?query= parameter to send it a SQL query string, so we can run anything.

grcevski replied:
That is true, but the test isn't really verifying if this worked if I'm not mistaken? It seems to only check for SELECT being there.

MattFrick commented on test/oats/configs/agent-config.river:
```diff
@@ -0,0 +1,47 @@
+logging {
```
The merge from main didn't remove these files, I am cleaning them out...

MattFrick replied:
Cherrypicked the two commits from earlier to be consistent.

MattFrick commented:
Sorry for the commit noise.  I had merged main into the branch but that still required rebasing, so I started over and rebased to main to keep things more in order.

grcevski commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,104 @@
+    struct sql_request_t sql_request = {0};
+    sql_request.start_time = bpf_ktime_get_ns();
+
+    if (should_include_db_statement) {
```
I'm not sure what happened to my comment here. I believe the more optimal approach is to do all the reading at the return of the queryDC. There are couple of reasons for that:

- We don't need an additional buffer, we can read directly into the trace event buffer that's allocated from the ringbuf.
- We can reuse our function invocation data structure, e.g. no need for another struct to be maintaned.

This approach is also different than OTel and we don't have to carry the copyright notice for code that's mostly written by Grafana. If we did this I don't think we need to copy any code from Otel.

MattFrick replied:
OK I've refactored all otel code out, and made it more consistent with the other probes by copying the full set of registers into an ongoing transaction map and using a few of them in the return probe.

grcevski commented on bpf/go_sql.c:
```diff
@@ -0,0 +1,104 @@
+    struct sql_request_t sql_request = {0};
+    sql_request.start_time = bpf_ktime_get_ns();
+
+    if (should_include_db_statement) {
```
When would we not want to send the sql event? We prune it before it's sent to traces, so there's no problem this being sent as is always. There's one less option and code path that needs to be tested.

MattFrick replied:
Config has been removed.

grcevski commented on docs/sources/configure/options.md:
```diff
@@ -169,6 +169,12 @@ can help with reducing the CPU overhead of Beyla.
 add a noticeable delay in the time the metrics are submitted and become externally visible.
 
+| YAML                       | Env var                    | Type | Default |
+| -------------------------- | -------------------------- | ---- | ------- |
```
We shouldn't need this option.

MattFrick replied:
Config has been removed.

grcevski commented on pkg/internal/ebpf/common/spanner.go:
```diff
@@ -95,3 +101,45 @@ func extractTraceparent(traceparent [55]byte) string {
+	}
+	switch operation {
+	case "ALTER", "CREATE", "DROP", "TRUNCATE":
+		if strings.ToUpper(fields[1]) == "TABLE" {
```
I think we can fix the cycops by extracting in another function, if this code is needed it shouldn't be commented out.

MattFrick replied:
This has been refactored to use new sqlprune package.

grcevski commented on pkg/internal/ebpf/common/spanner.go:
```diff
@@ -95,3 +101,45 @@ func extractTraceparent(traceparent [55]byte) string {
+		return operation, ""
+	}
+	table := ""
+	if operation == "UPDATE" {
```
I don't think this would work always, there can be operators for UPDATE, e.g. UPDATE TOP ... PERCENT...

MattFrick replied:
This has been refactored to use new sqlprune package.

grcevski commented on pkg/internal/ebpf/common/spanner.go:
```diff
@@ -95,3 +101,45 @@ func extractTraceparent(traceparent [55]byte) string {
+	if len(fields) < 3 {
+		return operation, ""
+	}
+	switch operation {
```
This needs to be a bit more complex, we need to skip tokens until we find one of the ones we are looking for. For example, 

WITH, UNION

grcevski replied:
Honestly, looking at how brittle this parsing might be and all the edge cases, did you perhaps consider in your design using a library to parse the SQL statement? Was there a performance or dependency reason why not to use something like https://github.com/xwb1989/sqlparser?

MattFrick replied:
That parser was my first plan, but I was told that performance-wise, we do not want to do a full parse on each operation, that we should try to do something simple to get the operation and table that works in most (but not necessarily all) cases.  I don't think we're going to end up with something good if we make our own parser that goes beyond the simplest case such as only finding the first word and assuming that it is actually something useful, such as the db operation.  Perhaps we may need to end up using a real parser.  The config option that controls copying the DB statement could disable it if the performance is unacceptable.  We'd still get the DB spans/metrics, just without the operation/table name.  One downside would be if the db query statement is truncated due to limited buffer size in ringbuf and the parse fails, it may just waste a lot of time.
What do you think, is it likely that using xwb1989 would create too much of a performance impact?  Would it be overly complicated to have a tri-state config: 1) use full sql parser, 2) use first word only, 3) don't process db statement at all.

grcevski replied:
Yeah, that makes sense about the performance, thanks! Let me answer in order what my thoughts are: 

> The config option that controls copying the DB statement could disable it if the performance is unacceptable. We'd still get the DB spans/metrics, just without the operation/table name.

I don't think this is an issue, we are doing application observability with Beyla and the overhead is expected to be higher, copying buffers overhead should be minimal here. We already do that for HTTP/gRPC. In a sense, number of queries without additional details can be determined with Infrastructure O11y, where App O11y does better/more is getting the level of detail that helps the developer.

> What do you think, is it likely that using xwb1989 would create too much of a performance impact? 

I think the only way to find out is to make a small benchmark to see how much slower is it on some sample queries, but I think the other point you made is stronger against this suggestion. If the parser fails completely on an incomplete query, without returning anything, then it's probably not important if it's fast enough. Does the parser fail without returning even a table name and operation if the query string is truncated?

If the parser is not an option, I think the code can be changed to loop a bit until it finds what it wants. Given that in 99% of cases it will be only one loop iteration, it shouldn't matter for performance. I think TOP would be common in MSSQL, but not elsewhere. But skipping WITH and RECURSIVE and other keywords until we find what the correct op is should be done. It's maybe worthwhile looking at docs in few SQL engines to see all the traps and building something more robust.

MattFrick replied:
I'm working on a new approach: Use xwb1989 tokenizer Scan() to find the type of each next token.  Find DB operation by matching against the set of db operation types, and find the (presumed) table name finding the value of the first "ID" type token that it finds.
I think this should be a reasonable balance: we don't fully parse it (so we don't end up failing when something is truncated,) we don't directly handle the tokens and all the nuance ourselves, we can reliable determine if something is a operation or a name, however it may be a wrong assumption that the first found "ID" type token would be a table name, although that should not be a sensitive token, so I think the risk is very low of leaking sensitive data values.

grcevski commented:
This PR is starting to look very goo. I think some of the conversations have been lost, I just wanted to re-iterate that we don't want to be reading twice for the SQL buffer, it should happen once in the 'returns' uprobe.

grcevski reviewed (commented):
Hey Matt, have you seen these tests here, I think they are very good to test that we can correctly process the majority of SQL statements.

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api-semconv/src/test/java/io/opentelemetry/instrumentation/api/db/SqlStatementSanitizerTest.java

MattFrick commented:
Thanks Nikola. I've added all those tests from the otel java sanitizer. I've also added a set of parameterized queries.  I'm feeling more confident that we'll have the correct db operation and table name.

MrAlias commented on bpf/go_sql.c:
Can you please provide appropriate attribution if you plan to copy work from OpenTelemetry.

MattFrick replied:
Absolutely.  Added note of inspiration and permalink to go sql probe.  Thanks!

pellared replied:
This is not an appropriate attribution. You must not change the copywrites. 

> // Copyright The OpenTelemetry Authors

See https://github.com/keyval-dev/opentelemetry-go-instrumentation/blob/ff642a6bcbcb83799c61d4631d07dcb9f50af5b6/LICENSE#L89-L128

grcevski replied:
The current version of the code is not a derivative work anymore, some early versions were, however we will add the attribution. There's an issue opened about this here https://github.com/grafana/beyla/issues/362.

grcevski commented on pkg/internal/export/otel/metrics.go:
```diff
@@ -322,6 +329,10 @@ func (mr *MetricsReporter) metricAttributes(span *request.Span) attribute.Set {
 			attrs = append(attrs, semconv.NetSockPeerName(span.Host))
 			attrs = append(attrs, semconv.NetSockPeerPort(span.HostPort))
 		}
+	case request.EventTypeSQLClient:
```
Sorry I didn't notice this before, but I think we should not use the `span.Path`, this will lead to cardinality explosion. What we are setting is a table name, which on certain systems can change often, e.g. folks may have daily tables, hourly tables sometimes.

I think it's best we put the DBOperation instead, which is much lower cardinality, in your case `span.Method`. Along these lines, I also don't think the table name would be very useful without the operation.

grcevski commented on pkg/internal/export/otel/traces.go:
```diff
@@ -321,6 +321,18 @@ func (r *TracesReporter) traceAttributes(span *request.Span) []attribute.KeyValu
+			table := span.Path
+			if table != "" {
+				attrs = append(attrs, semconv.DBSQLTable(table))
+				attrs = append(attrs, semconv.DBStatement(operation+" "+table))
```
Since we don't actually have the full statement, we shouldn't set this attribute DBStatement. It adds extra data and it contains no new value we already haven't set. According to the OTel spec:

```
// DBStatementKey is the attribute Key conforming to the "db.statement"
// semantic conventions. It represents the database statement being
// executed.
```

grcevski commented on pkg/internal/export/prom/prom.go:
```diff
@@ -36,6 +37,8 @@ const (
 	rpcMethodKey         = "rpc_method"
 	rpcSystemGRPC        = "rpc_system"
+	DBOperationKey       = "db_operation"
+	DBSQLTableTable      = "db_sql_table"
```
Similar here, we shouldn't put the DB table in the metrics for now. Let's just stick with operation, since it's guaranteed to be low cardinality if we parse things correctly.

I'll leave it up to you if you think we should put the table name, it can be under an option like we have for some of the HTTP attributes. If you choose to keep it with an option. it should be added as well for the OTel metrics (e.g. Prometheus and OTel should export the same things).

Honestly, I think we should leave it out for now and add it later if needed.

grcevski commented on pkg/internal/sqlprune/xwbparameterized_test.go:
```diff
@@ -0,0 +1,1081 @@
+/*
+Copyright 2017 Google Inc.
```
Why do we need these tests in our repo pkg folder? I think they should be in vendor? DId you modify the  xwb tests?

MattFrick replied:
We're using xwb's Scan() to lex, but we have our own simplistic parser that extracts the operation and table name based on the type of tokens that were found.  I have two sets of this: the original (w/ changed package name,) and the parameterized version where many of the ID tokens have been changed to question mark placeholders.  I expect our parser is going to see parameterized queries as the majority of cases.  Given that it does provide useful exercise to the sqlprune parser, it's only present in _test.go files, I'm not sure of the downside of leaving this stuff here - what do you think, should it still be removed?

grcevski replied:
Ah I see, thanks for explaining, keep them in then, no special downside except test time, but these are fast anyway.

grcevski reviewed (approved):
LGTM!

I'm approving assuming my latest feedback will be addressed:

- The metrics export should be standardized between OTel and Prometheus and contain only the operation for now.
- The duplicate DBStatement attribute should be removed from traces, since it doesn't add any new information and costs storage and CPU.
- The xwb parser tests that were copied from the vendor are testing the vendor code and they should not be part of our test suite. Since we now use the vendor to parse the SQL statement, rather than have our own code, we can rely on the parse testing that the vendor performs. We can still keep some of our basic tests to ensure most common queries still work, in case the vendor interface changes in the future.

mariomac commented:
Into another PR, I'd add some new integration tests. Coverage dropped from ~80% to 72% and seems that SQL traces are lacking integration tests. E.g. https://app.codecov.io/gh/grafana/beyla/tree/main/pkg%2Finternal%2Fsqlprune
